README.md
# koa-sleepy

[![NPM Version][npm-image]][npm-url]
[![Build Status][travis-image]][travis-url]
[![Node.js Version][node-version-image]][node-version-url]

router for building RESTful api

```js
const Koa = require('koa')
const {restify} = require('koa-sleepy')

class User {
    list (ctx, id) {}
    create (ctx) {}
    retrive (ctx, id) {}
    update (ctx, id) {}
    replace (ctx, id) {}
    destory (ctx, id) {}
    more(ctx, id) {}
}
const app = new Koa()
app.use(restify('/users', User))
app.listen()
```

to register multiple resources

```js
app.use(restify({
    '/users': User,
    '/comments': Comment
}))
```

[npm-image]: https://img.shields.io/npm/v/koa-sleepy.svg?style=flat
[npm-url]: https://npmjs.org/package/koa-sleepy
[travis-image]: https://img.shields.io/travis/zweifisch/koa-sleepy.svg?style=flat
[travis-url]: https://travis-ci.org/zweifisch/koa-sleepy
[node-version-image]: https://img.shields.io/node/v/koa-sleepy.svg
[node-version-url]: https://nodejs.org/en/download/


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

<a name="0.1.1-alpha.0"></a>
## 0.1.1-alpha.0 (2018-08-22)


test.js

const test = require('tape')
const resource = require('./index')


test('dispatch', ({equal, end}) => {
    equal()
    end()
})


test/sync.js
class Sync {

    create(ctx) {
    }

    retrive(ctx, id) {
        return id
    }

    replace(ctx, id) {
        ctx.body = {updated: id}
    }

    update(ctx, id) {
        return `${id} updated`
    }

    list(ctx) {
        return []
    }

    destroy(ctx, id) {
        return `${id} deleted`
    }

    custom(ctx, id) {
        return `custom action on ${id}`
    }
}

module.exports = Sync


test/async.js
const { BadRequest } = require('../index')

class Async {
    async list() {
        return []
    }

    async create() {
        return BadRequest({required: ["name"]})
    }
}

module.exports = Async


lib/response.js
class Response {
    constructor(status, message) {
        this.status = status
        this.message = message
    }
}

const response = code => message => new Response(code, message)

exports.Response = Response

exports.Ok = response(200)
exports.Created = response(201)
exports.Accepted = response(202)
exports.NoContent = response(204)
exports.BadRequest = response(400)
exports.Unauthorized = response(401)
exports.PaymentRequired = response(402)
exports.Forbidden = response(403)
exports.NotFound = response(404)
exports.MethodNotAllowed = response(405)
exports.NotAcceptable = response(406)
exports.ProxyAuthenticationRequired = response(407)
exports.RequestTimeout = response(408)
exports.Conflict = response(409)
exports.Gone = response(410)
exports.LengthRequired = response(411)
exports.PreconditionFailed = response(412)
exports.PayloadTooLarge = response(413)
exports.URITooLong = response(414)
exports.UnsupportedMediaType = response(415)
exports.RangeNotSatisfiable = response(416)
exports.ExpectationFailed = response(417)
exports.ImATeapot = response(418)
exports.MisdirectedRequest = response(421)
exports.UnprocessableEntity = response(422)
exports.Locked = response(423)
exports.FailedDependency = response(424)
exports.UnorderedCollection = response(425)
exports.UpgradeRequired = response(426)
exports.PreconditionRequired = response(428)
exports.TooManyRequests = response(429)
exports.RequestHeaderFieldsTooLarge = response(431)
exports.UnavailableForLegalReasons = response(451)
exports.InternalServerError = response(500)
exports.NotImplemented = response(501)
exports.BadGateway = response(502)
exports.ServiceUnavailable = response(503)
exports.GatewayTimeout = response(504)
exports.HTTPVersionNotSupported = response(505)
exports.VariantAlsoNegotiates = response(506)
exports.InsufficientStorage = response(507)
exports.LoopDetected = response(508)
exports.BandwidthLimitExceeded = response(509)
exports.NotExtended = response(510)
exports.NetworkAuthenticationRequired = response(511)


lib/restify.js
const assert = require('assert')
const debug = require('debug')('koa-sleepy')
const {Response} = require('./response')
const dispatch = require('./dispatch')
const compose = require('koa-compose')


const status = code => ctx => {
    ctx.status = code
}

const defaultHandlers = {
    methodNotAllowed: status(405),
    notFund: status(404)
}

function restify (...args) {
    if (args.length === 1) {
        assert(args[0] && typeof args[0] === 'object', `restify expact an object`)
        return compose(Object.entries(args[0]).map(x => restifySingle(...x)))
    }
    if (args.length === 2) {
        return restifySingle(...args)
    }
}

function restifySingle (prefix, resourceClass) {
    assert(typeof prefix === 'string', `restify expect prefix of string`)
    assert(resourceClass, `can't restify ${resourceClass}`)
    const controller = Reflect.construct(resourceClass, [])
    let actions = new Set(Object.getOwnPropertyNames(Object.getPrototypeOf(controller)))
    actions.delete('constructor')
    return async function(ctx, next) {
        if (!ctx.path.startsWith(prefix)) {
            await next()
            return
        }
        let [action, id] = dispatch(actions, ctx.method, ctx.path.substr(prefix.length))
        debug('dispatched', {method: ctx.method, path: ctx.path, action, id})
        if (actions.has(action)) {
            let result
            try {
                result = controller[action](ctx, id)
            } catch (err) {
                return ctx.throw(500, err)
            }
            if (result === undefined) {
                if (!ctx.body) {
                    ctx.status = 204
                }
            } else {
                if (result && 'function' === typeof result.then) {
                    try {
                        result = await result
                    } catch (err) {
                        ctx.throw(500, err)
                    }
                }
                if (result instanceof Response) {
                    ctx.status = result.status
                    ctx.body = result.message
                } else {
                    ctx.body = result
                }
            }
        } else {
            defaultHandlers[action](ctx, id)
        }
    }
}

exports.restify = restify
